sql/src/main/java/io/crate/operation/projectors/MergeProjector.java

```diff
@@ -0,0 +1,258 @@
+import io.crate.core.collections.Buckets;
+import io.crate.core.collections.Row;
+import io.crate.core.collections.RowN;
+import io.crate.operation.RowDownstream;
+import io.crate.operation.RowDownstreamHandle;
+import io.crate.operation.RowUpstream;
+import io.crate.operation.projectors.sorting.OrderingByPosition;
+import org.elasticsearch.common.Nullable;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class MergeProjector implements Projector  {
+
+    private final Ordering<Row> ordering;
+    private List<MergeProjectorDownstreamHandle> downstreamHandles = new ArrayList<>();
+    private final AtomicReference<Throwable> upstreamFailure = new AtomicReference<>(null);
+
+    private RowDownstreamHandle downstreamContext;
+
+    private Row lowestToEmit = null;
+
+    private boolean downstreamAborted = false;
+
+    public MergeProjector(int[] orderBy,
+                          boolean[] reverseFlags,
+                          Boolean[] nullsFirst) {
+        List<Comparator<Row>> comparators = new ArrayList<>(orderBy.length);
+        for (int i = 0; i < orderBy.length; i++) {
+            comparators.add(OrderingByPosition.rowOrdering(orderBy[i], reverseFlags[i], nullsFirst[i]));
+        }
+        ordering = Ordering.compound(comparators);
+    }
+
+    @Override
+    public void startProjection() {
+        if (downstreamHandles.size() <= 0) {
+            finishDownStream();
+        }
+    }
+
+    @Override
+    public RowDownstreamHandle registerUpstream(RowUpstream upstream) {
+        MergeProjectorDownstreamHandle handle = new MergeProjectorDownstreamHandle(this);
+        downstreamHandles.add(handle);
+        return handle;
+    }
+
+    @Override
+    public void downstream(RowDownstream downstream) {
+        downstreamContext = downstream.registerUpstream(this);
+    }
+
+    public void upstreamFinished(MergeProjectorDownstreamHandle handle) {
+        if (!handle.isEmpty()) {
+            return;
+        }
+        synchronized (this) {
+            downstreamHandles.remove(handle);
+        }
+        if (!downstreamAborted) {
+            emit();
+        }
+        finishDownStream();
+    }
+
+    public void finishDownStream() {
+        if (downstreamHandles.size() > 0) {
+            return;
+        }
+        if (downstreamContext != null) {
+            Throwable throwable = upstreamFailure.get();
+            if (throwable == null) {
+                downstreamContext.finish();
+            } else {
+                downstreamContext.fail(throwable);
+            }
+        }
+    }
+
+    public void upstreamFailed(Throwable throwable, MergeProjectorDownstreamHandle handle) {
+        upstreamFailure.set(throwable);
+        synchronized (this) {
+            downstreamHandles.remove(handle);
+        }
+        downstreamAborted();
+    }
+
+    private void downstreamAborted() {
+        downstreamAborted = true;
+        if (downstreamContext != null) {
```

Comment:
after reading this, would recommend AtomicBoolean for 'downstreamAborted' and use compareAndSet here to prevent double closing of downstream handles